cmd/devp2p/crawlercmd.go

```diff
@@ -0,0 +1,388 @@
+}
+
+func discv4(ctx *cli.Context, inputSet nodeSet, timeout time.Duration) nodeSet {
+	disc := startV4(ctx)
+	defer disc.Close()
+	// Crawl the DHT for some time
+	c := newCrawler(inputSet, disc, disc.RandomNodes())
+	c.revalidateInterval = 10 * time.Minute
+	return c.run(timeout)
+}
+
+// makeGenesis is the pendant to utils.MakeGenesis
+// with local flags instead of global flags.
+func makeGenesis(ctx *cli.Context) *core.Genesis {
+	switch {
+	case ctx.Bool(utils.RopstenFlag.Name):
+		return core.DefaultRopstenGenesisBlock()
+	case ctx.Bool(utils.RinkebyFlag.Name):
+		return core.DefaultRinkebyGenesisBlock()
+	case ctx.Bool(utils.GoerliFlag.Name):
+		return core.DefaultGoerliGenesisBlock()
+	case ctx.Bool(utils.YoloV3Flag.Name):
+		return core.DefaultYoloV3GenesisBlock()
+	default:
+		return core.DefaultGenesisBlock()
+	}
+}
+
+func crawlRound(ctx *cli.Context, inputSet nodeSet, influxdb *influx, timeout time.Duration) nodeSet {
+	output := make(nodeSet)
+	log.Info("DiscV5")
+	v5 := discv5(ctx, nodeSet{}, timeout)
+	output.add(v5.nodes()...)
+	log.Info("DiscV4")
+	v4 := discv4(ctx, nodeSet{}, timeout)
+	output.add(v4.nodes()...)
+
+	genesis := makeGenesis(ctx)
+	if genesis == nil {
+		genesis = core.DefaultGenesisBlock()
+	}
+	networkID := ctx.Uint64(utils.NetworkIdFlag.Name)
+	nodeURL := ctx.String(nodeURLFlag.Name)
+
+	reqChan := make(chan nodeJSON, len(output))
+	respChan := make(chan crawledNode, 10)
+	getNodeLoop := func(in <-chan nodeJSON, out chan<- crawledNode) {
+		for {
+			node := <-in
+			info, err := getClientInfo(genesis, networkID, nodeURL, node.N)
+			if err != nil {
+				log.Warn("GetClientInfo failed", "error", err, "nodeID", node.N.ID())
+			} else {
+				log.Info("GetClientInfo succeeded")
+			}
+			out <- crawledNode{node: node, info: info}
+		}
+	}
+	// Schedule 10 workers
+	for i := 0; i < 10; i++ {
+		go getNodeLoop(reqChan, respChan)
+	}
+
+	// Try to connect and get the status of all nodes
+	for _, node := range output {
+		reqChan <- node
+	}
+	var nodes []crawledNode
+	for i := 0; i < len(output); i++ {
+		node := <-respChan
+		nodes = append(nodes, node)
+	}
+	// Write the node info to influx
+	if influxdb != nil {
+		if err := influxdb.updateNodes(nodes); err != nil {
+			exit(err)
+		}
+	}
+	return output
+}
+
+func getClientInfo(genesis *core.Genesis, networkID uint64, nodeURL string, n *enode.Node) (*clientInfo, error) {
+	var info clientInfo
+	conn, sk, err := dial(n)
+	if err != nil {
+		return nil, err
+	}
+	defer conn.Close()
+	conn.SetDeadline(time.Now().Add(5 * time.Second))
+
+	// write hello to client
+	pub0 := crypto.FromECDSAPub(&sk.PublicKey)[1:]
+	ourHandshake := &ethtest.Hello{
+		Version: 5,
+		Caps: []p2p.Cap{
+			{Name: "eth", Version: 64},
+			{Name: "eth", Version: 65},
+			{Name: "eth", Version: 66},
+		},
+		ID: pub0,
+	}
+	if err := conn.Write(ourHandshake); err != nil {
+		return nil, err
+	}
+
+	// read hello from client
+	switch msg := conn.Read().(type) {
+	case *ethtest.Hello:
```

Comment:
Nitpick, not sure it's proper to define the message type in the "test" package. Perhaps in the future we should move them to a better place?